tools/points-service/main.go

```diff
@@ -0,0 +1,529 @@
+	for rows.Next() {
+		var id int64
+		var inBlock uint64
+		if scanErr := rows.Scan(&id, &inBlock); scanErr != nil {
+			logger.Error("scan error", "error", scanErr)
+			continue
+		}
+
+		var blocksActive int64
+		if currentBlock > inBlock {
+			blocksActive = int64(currentBlock - inBlock)
+		}
+		totalPoints := computePointsForMonths(blocksActive)
+
+		_, updErr := tx.Exec(`
+            UPDATE events
+            SET points_accumulated = ?
+            WHERE id = ?
+        `, totalPoints, id)
+		if updErr != nil {
+			logger.Error("failed to update points", "error", updErr, "id", id)
+		}
+	}
+	if rowsErr := rows.Err(); rowsErr != nil {
+		err = fmt.Errorf("rows iteration error: %w", rowsErr)
+	}
+
+	return err
+}
+
+// StartPointsRoutine spawns a background goroutine that runs every 'interval'
+// and calls updatePoints with the latest block.
+func StartPointsRoutine(db *sql.DB, logger *slog.Logger, interval time.Duration, ethClient *ethclient.Client) {
+	// Run once immediately
+	logger.Info("Starting initial points accrual run")
+	latestBlock, err := ethClient.BlockByNumber(context.Background(), nil)
+	if err != nil {
+		logger.Error("cannot fetch latest block", "error", err)
+	} else {
+		currBlockNum := latestBlock.NumberU64()
+		if err := updatePoints(db, logger, currBlockNum); err != nil {
+			logger.Error("initial points accrual run failed", "error", err)
+		} else {
+			logger.Info("initial points accrual run completed successfully")
+		}
+	}
+
+	ticker := time.NewTicker(interval)
+	go func() {
+		defer ticker.Stop()
+
+		for range ticker.C {
+			logger.Info("Starting points accrual run")
+
+			// fetch latest block
+			latestBlock, err := ethClient.BlockByNumber(context.Background(), nil)
+			if err != nil {
+				logger.Error("cannot fetch latest block", "error", err)
+				continue
+			}
+			currBlockNum := latestBlock.NumberU64()
+
+			// do the update inside a transaction
+			if err := updatePoints(db, logger, currBlockNum); err != nil {
+				logger.Error("points accrual run failed", "error", err)
+			} else {
+				logger.Info("points accrual run completed successfully")
+			}
+		}
+	}()
+}
+
+// --------------------------------------
+//   3) SERVICE + EVENT HANDLERS
+// --------------------------------------
+
+type PointsService struct {
+	logger    *slog.Logger
+	db        *sql.DB
+	ethClient *ethclient.Client
+	block     uint64
+}
+
+func (ps *PointsService) LastBlock() (uint64, error) {
+	return ps.block, nil
+}
+func (ps *PointsService) SetLastBlock(block uint64) error {
+	ps.block = block
+	return nil
+}
+
+// Insert new row for opt-in
+func insertOptIn(db *sql.DB, logger *slog.Logger, pubkey, adder, registryType, eventType string, inBlock uint64) {
+	rwLock.RLock()
+	defer rwLock.RUnlock()
+
+	var existingAdder string
+	err := db.QueryRow(`
+        SELECT adder FROM events
+        WHERE pubkey = ? AND opted_out_block IS NULL
+    `, pubkey).Scan(&existingAdder)
+
+	// if pubkey is active under a different adder, ignore
```

Comment:
This can be avoided by just checking the error on Insert. I think sql should give you a specific error for unique key violation.

Reply:
The reason we have this because teh adder pubkey is not unqiue, it's adder, piubkey and opted in block, therefore sql won't throw an error